# RAG(Retrieval Augmented Generation) use local model, production grade RAG.


### 1. import the libraries

In [63]:
import os
import re
from dotenv import load_dotenv
load_dotenv()
from pathlib import Path
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_huggingface import HuggingFaceEmbeddings
from langchain_chroma import Chroma
from langchain_core.documents import Document
import pymupdf4llm
from langchain_text_splitters import MarkdownHeaderTextSplitter


### 2. Convert PDF file to markdown

In [64]:
# Convert the pdf to markdown using pymupdf4llm(replace pymupdfloader from langchain to solve the section context issue for one of the eval test set - that only has half the answer when retrieved))
md_text = pymupdf4llm.to_markdown("data/Claude Certified Architect – Foundations Certification Exam Guide.pdf")

# Clean the markdown text by removing extra newlines and spaces
md_text = re.sub(r"^\s*-{3,}\s*$","",md_text,flags=re.MULTILINE)

md_text = re.sub(r"\n{3,}","\n\n",md_text)
print(md_text)


Processing data/Claude Certified Architect – Foundations Certification Exam Guide.pdf...
[                                        ] (0/4[=                                       ] ( 1/4[==                                      ] ( 2/40[===                                     ] ( 3/4[====                                    ] ( 4/40[=====                                   ] ( 5/4[======                                  ] ( 6/40[=======                                 ] ( 7/4[========                                ] ( 8/40[=========                               ] ( 9/4[==========                              ] (10/40[===========                             ] (11/4[============                            ] (12/40[=============                           ] (13/4[==============                          ] (14/40[===============                         ] (15/4[================                        ] (16/40[=================                       ] (17/4[==================                     

In [65]:
print(f"Length of markdown text: {len(md_text)} characters and {len(md_text.split())} words")

Length of markdown text: 77836 characters and 10519 words


### 3. Split the markdown into chunks base on headers use Markdown Hearder Splitter

In [66]:
# Split the markdown text into chunks based on headers
headers_to_split_on = [
    ("##", "H2"),
    ("###", "H3"),
]
md_splitter = MarkdownHeaderTextSplitter(headers_to_split_on=headers_to_split_on)
head_chunks = md_splitter.split_text(md_text)
print(f"Split markdown into {len(head_chunks)} chunks based on headers\n")

for i,chunk in enumerate(head_chunks):
    print(f"---chunk{i}----\n")
    print(f"metadata: {chunk.metadata}\n")
    print(f"{chunk.page_content}\n")
    print(f"Length of header chunks:{len(chunk.page_content)}\n")


Split markdown into 57 chunks based on headers

---chunk0----

metadata: {}

# Claude Certified Architect – Foundations Certification Exam Guide

Length of header chunks:67

---chunk1----

metadata: {'H2': 'Introduction'}

The Claude Certified Architect – Foundations certification validates that practitioners can make
informed decisions about tradeoffs when implementing real-world solutions with Claude. This
exam tests foundational knowledge across Claude Code, the Claude Agent SDK, the Claude API,
and Model Context Protocol (MCP) — the core technologies used to build production-grade
applications with Claude.  
Questions on this exam are grounded in realistic scenarios drawn from actual customer use
cases, including building agentic systems for customer support, designing multi-agent research
pipelines, integrating Claude Code into CI/CD workflows, building developer productivity tools,
and extracting structured data from unstructured documents. Candidates must demonstrate not
only co

### 4. Chunking strategies 
### fixed size, recursive text splitter, section chunk, sentence chunk, semantic chunk

#### 4.1 Apply Recursive Character Text Splitter on head_chunks with breadcrumb to maintain section context(this to fix the eval set that only appear half of the answer for Q&A)

In [67]:
# Further split the chunks into smaller pieces from the head_chunks and append the header info to the content use breadcrumb style (e.g. Section: H1 > H2 > H3) to preserve the context of the section for better retrieval results.
def chunk_recursive_with_breadcrumbs(head_chunks):
    text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
    md_chunks=text_splitter.split_documents(head_chunks)
    
    for chunk in md_chunks:
        breadcrumb = " > ".join(
            v for k, v in chunk.metadata.items() if k.lower().startswith("h")
        )
        chunk.page_content = f"Section: {breadcrumb}\n\n{chunk.page_content}"

    return md_chunks   


In [68]:
md_chunks_breadcrumb = chunk_recursive_with_breadcrumbs(head_chunks)

print(f"Split {len(head_chunks)} documents into {len(md_chunks_breadcrumb)} chunks\n")

for i, chunk in enumerate(md_chunks_breadcrumb[:9]):
    print(f"Chunk {i}:\n{chunk.page_content}\nLength of chunk:{len(chunk.page_content)}\n{'-'*50}\n")


Split 57 documents into 113 chunks

Chunk 0:
Section: 

# Claude Certified Architect – Foundations Certification Exam Guide
Length of chunk:78
--------------------------------------------------

Chunk 1:
Section: Introduction

The Claude Certified Architect – Foundations certification validates that practitioners can make
informed decisions about tradeoffs when implementing real-world solutions with Claude. This
exam tests foundational knowledge across Claude Code, the Claude Agent SDK, the Claude API,
and Model Context Protocol (MCP) — the core technologies used to build production-grade
applications with Claude.  
Questions on this exam are grounded in realistic scenarios drawn from actual customer use
cases, including building agentic systems for customer support, designing multi-agent research
pipelines, integrating Claude Code into CI/CD workflows, building developer productivity tools,
and extracting structured data from unstructured documents. Candidates must demonstrate not
onl

### 4.2 Chunks without breadcrumb using Recursive Character Text Splitter, with metadata in each chunk

In [69]:
# Chunking the documents using RecursiveCharacterTextSplitter from Langchain, it is a smarter variant of fixed_size chunking.It tries to cut at natural boundaries
def chunk_recursive(head_chunks, chunk_size=1000, chunk_overlap=200):
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=chunk_size,
        chunk_overlap=chunk_overlap,
        length_function=len,
        separators=["\n\n", "\n", ". "," ", ""] 
    )
    split_docs = text_splitter.split_documents(head_chunks)

    return split_docs

In [70]:
recursive_char = chunk_recursive(head_chunks)
print(f"Split {len(head_chunks)} documents into {len(recursive_char)} chunks with recursive character splitting")

for i, chunk in enumerate(recursive_char[:5]):
    print(f"--- Chunk {i} ---")
    print(f"Content: {chunk.page_content}\n")
    print(f"Metadata: {chunk.metadata}\n")
    print(f"Length of chunk: {len(chunk.page_content)} characters\n")
    

Split 57 documents into 113 chunks with recursive character splitting
--- Chunk 0 ---
Content: # Claude Certified Architect – Foundations Certification Exam Guide

Metadata: {}

Length of chunk: 67 characters

--- Chunk 1 ---
Content: The Claude Certified Architect – Foundations certification validates that practitioners can make
informed decisions about tradeoffs when implementing real-world solutions with Claude. This
exam tests foundational knowledge across Claude Code, the Claude Agent SDK, the Claude API,
and Model Context Protocol (MCP) — the core technologies used to build production-grade
applications with Claude.  
Questions on this exam are grounded in realistic scenarios drawn from actual customer use
cases, including building agentic systems for customer support, designing multi-agent research
pipelines, integrating Claude Code into CI/CD workflows, building developer productivity tools,
and extracting structured data from unstructured documents. Candidates must demonstrate

In [71]:
# Fixed-size chunking by character count with overlap. This way the chunks may not make sense because it cuts in the middle of sentences.
def chunks_by_char(md_chunks, chunk_size=1000, chunk_overlap=200):
    chunks = []

    for doc in md_chunks:
            page = doc.page_content
            start_idx = 0

            while start_idx < len(page):
                end_idx = min(start_idx + chunk_size, len(page))
                chunk = page[start_idx:end_idx]
                chunks.append(chunk)

                start_idx = (
                end_idx - chunk_overlap if end_idx < len(page) else end_idx
             )

    return chunks   

In [72]:
all_pdf_chunks = chunks_by_char(head_chunks)
print(f"Split {len(head_chunks)} documents into {len(all_pdf_chunks)} character chunks with overlap")
for i, chunk in enumerate(all_pdf_chunks[:5]):
    print(f"--- Chunk{i} ---\n{chunk}\n")

Split 57 documents into 113 character chunks with overlap
--- Chunk0 ---
# Claude Certified Architect – Foundations Certification Exam Guide

--- Chunk1 ---
The Claude Certified Architect – Foundations certification validates that practitioners can make
informed decisions about tradeoffs when implementing real-world solutions with Claude. This
exam tests foundational knowledge across Claude Code, the Claude Agent SDK, the Claude API,
and Model Context Protocol (MCP) — the core technologies used to build production-grade
applications with Claude.  
Questions on this exam are grounded in realistic scenarios drawn from actual customer use
cases, including building agentic systems for customer support, designing multi-agent research
pipelines, integrating Claude Code into CI/CD workflows, building developer productivity tools,
and extracting structured data from unstructured documents. Candidates must demonstrate not
only conceptual knowledge but practical judgment about architecture, conf

In [73]:
def chunks_by_sentences(head_chunks, max_sentences_per_chunk=5, overlap_sentences=1):
    chunks = []

    for doc in head_chunks:
        page = doc.page_content
        sentences = re.split(r"(?<=[.?!])\s+",page)

        start_idx = 0
        while start_idx < len(sentences):
            end_idx = min(start_idx + max_sentences_per_chunk, len(sentences))
            current_chunk = sentences[start_idx:end_idx]
            chunks.append(" ".join(current_chunk))
            start_idx = end_idx - overlap_sentences if end_idx < len(sentences) else end_idx

    return chunks


In [74]:
all_pdf_sentence_chunks = chunks_by_sentences(head_chunks)
print(f"Split {len(head_chunks)} documents into {len(all_pdf_sentence_chunks)} sentence chunks")
for i, chunk in enumerate(all_pdf_sentence_chunks[:5]):
    print(f"--- Sentence Chunk {i} ---\n{chunk}\n")

Split 57 documents into 104 sentence chunks
--- Sentence Chunk 0 ---
# Claude Certified Architect – Foundations Certification Exam Guide

--- Sentence Chunk 1 ---
The Claude Certified Architect – Foundations certification validates that practitioners can make
informed decisions about tradeoffs when implementing real-world solutions with Claude. This
exam tests foundational knowledge across Claude Code, the Claude Agent SDK, the Claude API,
and Model Context Protocol (MCP) — the core technologies used to build production-grade
applications with Claude. Questions on this exam are grounded in realistic scenarios drawn from actual customer use
cases, including building agentic systems for customer support, designing multi-agent research
pipelines, integrating Claude Code into CI/CD workflows, building developer productivity tools,
and extracting structured data from unstructured documents. Candidates must demonstrate not
only conceptual knowledge but practical judgment about architecture, 

##### For chunking, will proceed with langchain recursive character text splitter with breadcrumb where each section has joint header which solve the section context issue in eval test(Only keep the header H2&h3 for each section , not H1 to reduce noise"Exam Guide" to boost rank up). 
##### Fixed size with overlap splitter has hard stop and break the words which causing missing info/lack context.
##### Chunk by sentences leave the chunk sizes varies in different size which is not consistent for token ceiling.
##### Chunk by section in PDF is chunking by pages not section - it works the best on markdown files.
##### Will try semantic chunking later which is more expensive because use LLM to evaluate.

## 3. Embeddings & Vector Database

In [75]:
#sentence-transformers/all-MiniLM-L6-v2  produces a 384-dimensional vector for each chunk. 
def create_embeddings(documents):
    embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")
    if Path("./chroma_db").exists(): # load existing vector store
        vector_store = Chroma(collection_name="md_chunks_breadcrumb", embedding_function=embedding_model, persist_directory="./chroma_db")
    else:
        vector_store = Chroma.from_documents(md_chunks_breadcrumb, embedding_model, collection_name="md_chunks_breadcrumb",persist_directory="./chroma_db")
    return vector_store

In [76]:
vectorstore=create_embeddings(md_chunks_breadcrumb)

KeyboardInterrupt: 

In [ ]:
print(vectorstore._collection.count())

113


### 3.1 Add ground_truth_chunk_ids to eval_testset.

In [ ]:
#Read the eval set from JSONL
import json
rows = []
with open("eval/eval_testset_v1.jsonl","r",encoding="utf-8") as f:
    for line in f:
        rows.append(json.loads(line))

print(len(rows))
print(rows[0])


60
{'id': 'Q01', 'type': 'definition', 'question': 'What does the Claude Certified Architect – Foundations certification validate about its holders?', 'answer': 'That practitioners can make informed decisions about tradeoffs when implementing real-world solutions with Claude.', 'section': 'Introduction', 'page': 1, 'source_quote': 'The Claude Certified Architect – Foundations certification validates that practitioners can make informed decisions about tradeoffs when implementing real-world solutions with Claude.', 'ground_truth_chunk_ids': ['9fdeaa6d-faba-4278-ba6f-27c99a86216d']}


In [ ]:
#Get all chunks (id + text) from ChromaDB
all_chunks = vectorstore.get(include=["embeddings", "documents", "metadatas"])
print(all_chunks.keys())

n = 2
for i in range(n):
    print(f"\n ----- Chunk {i} -----")
    print(f"ids: {all_chunks["ids"][i]}")
    print("document:",all_chunks["documents"][i])
    print("embeddings(first 10 dimensions):",all_chunks["embeddings"][i][:10])


dict_keys(['ids', 'embeddings', 'documents', 'uris', 'included', 'data', 'metadatas'])

 ----- Chunk 0 -----
ids: 9fdeaa6d-faba-4278-ba6f-27c99a86216d
document: Section: Introduction

The Claude Certified Architect – Foundations certification validates that practitioners can make
informed decisions about tradeoffs when implementing real-world solutions with Claude. This
exam tests foundational knowledge across Claude Code, the Claude Agent SDK, the Claude API,
and Model Context Protocol (MCP) — the core technologies used to build production-grade
applications with Claude.  
Questions on this exam are grounded in realistic scenarios drawn from actual customer use
cases, including building agentic systems for customer support, designing multi-agent research
pipelines, integrating Claude Code into CI/CD workflows, building developer productivity tools,
and extracting structured data from unstructured documents. Candidates must demonstrate not
only conceptual knowledge but practical judgme

In [ ]:
# Normalized substring check 

def normalize(text: str) -> str:
    text = re.sub(r"[*_#`]+", "",text)
    return " ".join(text.lower().split())

def chunk_contains_quote(chunk_text: str, quote: str) -> bool:
    return normalize(quote) in normalize(chunk_text)


In [ ]:
needs_review = []

for row in rows:
    if row["type"] == "no_answer":
        row["ground_truth_chunk_ids"] = []
        continue
    
    quote = row["source_quote"]
    matches = []
    
    for chunk_id, chunk_text in zip(all_chunks["ids"],all_chunks["documents"]):
        if chunk_contains_quote(chunk_text,quote):
           matches.append(chunk_id)
           
    if len(matches) == 0:
        needs_review.append(row)
    
    else:
        row["ground_truth_chunk_ids"] = matches


In [ ]:
with open("eval/eval_testset_v1.jsonl","w",encoding="utf-8") as f:
    for row in rows:
        f.write(json.dumps(row) + "\n")

with open("eval/needs_review.jsonl","w", encoding="utf-8") as f:
    for row in needs_review:
        f.write(json.dumps(row) + "\n")

print(f"Total rows: {len(rows)}")
print(f"Needs review:{len(needs_review)}")
print(f"Rows with ground truth:{sum(1 for r in rows if 'ground_truth_chunk_ids' in r and r['ground_truth_chunk_ids'])}")

Total rows: 60
Needs review:0
Rows with ground truth:48


In [ ]:
from transformers import AutoTokenizer

In [ ]:
tokenizer = AutoTokenizer.from_pretrained("sentence-transformers/all-MiniLM-L6-v2")


max_tokens = max(len(tokenizer.encode(c.page_content)) for c in md_chunks_breadcrumb)
print(f"Max tokens in the chunks: {max_tokens}")

Max tokens in the chunks: 278


## 4. Eval test set with similarity_search as baseline metrics

In [ ]:
# eval test with 15-20 sample query to see if the vector store is working
query1 = "What content domains and weightings do the exam have?"
results = vectorstore.similarity_search(query1, k=5)
for i, result in enumerate(results):
    print(f"\n--- Result {i} ---\n")
    print(result.page_content)
    print(result.metadata)


--- Result 0 ---

Section: Content Outline

This exam guide includes weightings, content domains, and task statements for the exam.  
The exam has the following content domains and weightings:  
- **Domain 1: Agentic Architecture & Orchestration (27% of scored content)**  
- **Domain 2: Tool Design & MCP Integration (18% of scored content)**  
- **Domain 3: Claude Code Configuration & Workflows (20% of scored content)**  
- **Domain 4: Prompt Engineering & Structured Output (20% of scored content)**  
- **Domain 5: Context Management & Reliability (15% of scored content)**
{'H2': 'Content Outline'}

--- Result 1 ---

Section: Appendix > Out-of-Scope Topics

The following related topics will NOT appear on the exam:  
- Fine-tuning Claude models or training custom models  
- Claude API authentication, billing, or account management  
- Detailed implementation of specific programming languages or frameworks (beyond
what's needed for tool and schema configuration)  
- Deploying or hosting

Query 1 retrieval came back with Result 0 which is correct.

In [ ]:
# query 2 to test the vector store
query2 = "What hands-on experience the idea candidate with for the exam?"
results = vectorstore.similarity_search(query2, k=30)
for i, result in enumerate(results):
    print(f"\n--- Result {i} ---\n")
    print(result.page_content)
    print(f"Length of result.page_content: {len(result.page_content)}")
Lengths = [len(result.page_content) for result in results]
print(f"min={min(Lengths)}, max ={max(Lengths)}, avg={sum(Lengths)/len(Lengths)}")
small = [result for result in results if len(result.page_content) < 150]
for r in small:
    print(repr(result.page_content[:200]))
    print("---")



--- Result 0 ---

Section: Appendix > Exam Preparation Recommendations

validation-retry loops, design schemas with optional/nullable fields, and practice batch
processing with the Message Batches API.  
5. Practice prompt engineering techniques: Write few-shot examples for ambiguous  
scenarios. Define explicit review criteria to reduce false positives. Design multi-pass
review architectures for large code reviews.  
6. Study context management patterns: Practice extracting structured facts from verbose  
tool outputs, implementing scratchpad files for long sessions, and designing subagent
delegation to manage context limits.  
7. Review escalation and human-in-the-loop patterns: Understand when to escalate (policy  
gaps, customer requests, inability to progress) versus resolve autonomously. Practice
designing human review workflows with confidence-based routing.  
8. Complete the Practice Exam: Before sitting for the real exam, complete the practice exam  
(the link will be provide

query2 didnot return any results even increased k=20. Check if the content is there use keyword search. Increased to k=30. Partial ranked 22. Also notice some of the results longer than 1000 chunk_size, min=78, max =1124, avg=706.9, max=1124 is 12% over target and expected.But min=78 will need to investigate, because it is way smaller.


In [ ]:
keyword="hands-on experience"
for i,chunk in enumerate(md_chunks_breadcrumb):
    if keyword.lower() in chunk.page_content.lower():
        print(f"Chunk {i} : {chunk.page_content}\n")
        

Chunk 3 : Section: Target Candidate Description

The ideal candidate for this certification is a solution architect who designs and implements
production applications with Claude. This candidate has hands-on experience with:  
- Building agentic applications using the Claude Agent SDK, including multi-agent
orchestration, subagent delegation, tool integration, and lifecycle hooks  
- Configuring and customizing Claude Code for team workflows using CLAUDE.md files,
Agent Skills, MCP server integrations, and plan mode  
- Designing Model Context Protocol (MCP) tool and resource interfaces for backend
system integration  
- Engineering prompts that produce reliable structured output, leveraging JSON schemas,
few-shot examples, and extraction patterns  
- Managing context windows effectively across long documents, multi-turn conversations,
and multi-agent handoffs  
- Integrating Claude into CI/CD pipelines for automated code review, test generation, and
pull request feedback



Note: 1. Content with keyword "hands-on experience" is there, the issue is retrieval.


In [ ]:
# test with query 3
query3 = "What are out of scope topics?"
results = vectorstore.similarity_search(query3, k=15)
for i, result in enumerate(results):
    print(f"\n--- Result {i} ---\n")
    print(result.page_content)
    



--- Result 0 ---

Section: Appendix > In-Scope Topics

fields to prevent hallucination  
- **Few-shot prompting: Ambiguous scenario targeting, format consistency, false positive**
reduction  
- **Batch processing: Message Batches API appropriateness, latency tolerance assessment,**  
failure handling by custom_id  
- **Context window optimization: Trimming verbose tool outputs, structured fact extraction,**
position-aware input ordering  
- **Human review workflows: Confidence calibration, stratified sampling, accuracy**
segmentation by document type and field  
- **Information provenance: Claim-source mappings, temporal data handling, conflict**
annotation, coverage gap reporting

--- Result 1 ---

Section: Exam Scenarios

The exam uses scenario-based questions. Each scenario presents a realistic production context
that frames a set of questions. During the exam, 4 scenarios will be presented and picked at
random from the full set of the 6 scenarios below.

--- Result 2 ---

Section:

Retrieval result for query 3: increased k=5 to 15 because only found result 7 has majority answer and even k=15 still not retrieve the other half. Next will try BM25 and rerun if this improves. And then hybrid or reranker.

In [ ]:
query4 = "What is the response type in this exam?"
results = vectorstore.similarity_search(query4, k=5)
for i, result in enumerate(results):
    print(f"\n --- Result {i} ---\n")
    print(result.page_content)
    print(result.metadata)
    


 --- Result 0 ---

Section: Exam Content > Response Types

All questions on the exam are multiple choice format. Each question has one correct response
and three incorrect responses (distractors).  
Select the single response that best completes the statement or answers the question.
Distractors are response options that a candidate with incomplete knowledge or experience
might choose.  
Unanswered questions are scored as incorrect; there is no penalty for guessing.
{'H2': 'Exam Content', 'H3': 'Response Types'}

 --- Result 1 ---

Section: Sample Questions

The following sample questions illustrate the format and difficulty level of the exam. These are
drawn from the practice test and include explanations to aid learning.
{'H2': 'Sample Questions'}

 --- Result 2 ---

Section: Exam Content > Exam Results

The exam has a pass or fail designation. The exam is scored against a minimum standard
established by subject matter experts.  
Your results are reported as a scaled score of 100–1,

Query 4: result 0 is covered the answer.

In [ ]:
query5 = "What will be the exam result looks like?"
results = vectorstore.similarity_search(query5, k=5)
for i, results in enumerate(results):
    print(f"\n --- Result {i} ---\n")
    print(results.page_content)
    print(results.metadata)



 --- Result 0 ---

Section: Exam Content > Exam Results

The exam has a pass or fail designation. The exam is scored against a minimum standard
established by subject matter experts.  
Your results are reported as a scaled score of 100–1,000. The minimum passing score is 720.
Scaled scoring models help equate scores across multiple exam forms that might have slightly
different difficulty levels.
{'H2': 'Exam Content', 'H3': 'Exam Results'}

 --- Result 1 ---

Section: Appendix > Exam Preparation Recommendations

8. Complete the Practice Exam: Before sitting for the real exam, complete the practice exam  
(the link will be provided separately). The practice exam covers the same scenarios and
question format as the real exam and shows explanations after each answer to help
reinforce your understanding.  
**Version 0.1** **Last Updated: Feb 10 2025**
{'H3': 'Exam Preparation Recommendations', 'H2': 'Appendix'}

 --- Result 2 ---

Section: Content Outline

This exam guide includes weighti

Query 5 retrieved result 0 which is correct

In [ ]:
query6 = "If I got 710, did I pass the exam?"
results = vectorstore.similarity_search(query6, k=5)
for i, results in enumerate(results):
    print(f"\n --- Result {i} ---\n")
    print(results.page_content)


 --- Result 0 ---

Section: Exam Content > Exam Results

The exam has a pass or fail designation. The exam is scored against a minimum standard
established by subject matter experts.  
Your results are reported as a scaled score of 100–1,000. The minimum passing score is 720.
Scaled scoring models help equate scores across multiple exam forms that might have slightly
different difficulty levels.

 --- Result 1 ---

Section: Appendix > Exam Preparation Recommendations

8. Complete the Practice Exam: Before sitting for the real exam, complete the practice exam  
(the link will be provided separately). The practice exam covers the same scenarios and
question format as the real exam and shows explanations after each answer to help
reinforce your understanding.  
**Version 0.1** **Last Updated: Feb 10 2025**

 --- Result 2 ---

Section: Exam Content > Response Types

All questions on the exam are multiple choice format. Each question has one correct response
and three incorrect responses (

query 6 retrieved Result 0 that cover the score. Is it correct to my question?

In [ ]:
query7 = "What topics are NOT covered in this exam?"
results = vectorstore.similarity_search(query7, k=10)
for i, results in enumerate(results):
    print(f"\n --- Result {i} ---\n")
    print(results.page_content)


 --- Result 0 ---

Section: Exam Scenarios

The exam uses scenario-based questions. Each scenario presents a realistic production context
that frames a set of questions. During the exam, 4 scenarios will be presented and picked at
random from the full set of the 6 scenarios below.

 --- Result 1 ---

Section: Sample Questions

The following sample questions illustrate the format and difficulty level of the exam. These are
drawn from the practice test and include explanations to aid learning.

 --- Result 2 ---

Section: Preparation Exercises

Complete these hands-on exercises to build practical familiarity with the topics covered on the
exam. Each exercise is designed to reinforce knowledge across one or more exam domains.

 --- Result 3 ---

Section: Content Outline

This exam guide includes weightings, content domains, and task statements for the exam.  
The exam has the following content domains and weightings:  
- **Domain 1: Agentic Architecture & Orchestration (27% of scored con

query7 returned result 6 that has 90% of answer - missing 2 points.

In [ ]:
query8 = "How long will the exam takes?"
results = vectorstore.similarity_search(query8, k=5)
for i, results in enumerate(results):
    print(f"\n --- Result {i} ---\n")
    print(results.page_content)


 --- Result 0 ---

Section: Appendix > Exam Preparation Recommendations

8. Complete the Practice Exam: Before sitting for the real exam, complete the practice exam  
(the link will be provided separately). The practice exam covers the same scenarios and
question format as the real exam and shows explanations after each answer to help
reinforce your understanding.  
**Version 0.1** **Last Updated: Feb 10 2025**

 --- Result 1 ---

Section: Exam Content > Exam Results

The exam has a pass or fail designation. The exam is scored against a minimum standard
established by subject matter experts.  
Your results are reported as a scaled score of 100–1,000. The minimum passing score is 720.
Scaled scoring models help equate scores across multiple exam forms that might have slightly
different difficulty levels.

 --- Result 2 ---

Section: Appendix > Exam Preparation Recommendations

validation-retry loops, design schemas with optional/nullable fields, and practice batch
processing with the M

In [ ]:
query9 = "What would happen if I used Plan mode instead of direct execution?"
results = vectorstore.similarity_search(query9, k=5)
for i, result in enumerate(results):
    print(f"\n --- Result {i} ---\n")
    print(result.page_content)



 --- Result 0 ---

Section: Domain 3: Claude Code Configuration & Workflows > Task Statement 3.4: Determine when to use plan mode vs direct execution

- Selecting direct execution for well-understood changes with clear scope (e.g., a single-file
bug fix with a clear stack trace, adding a date validation conditional)  
- Using the Explore subagent for verbose discovery phases to prevent context window
exhaustion during multi-phase tasks  
- Combining plan mode for investigation with direct execution for implementation (e.g.,
planning a library migration, then executing the planned approach)

 --- Result 1 ---

Section: Domain 3: Claude Code Configuration & Workflows > Task Statement 3.4: Determine when to use plan mode vs direct execution

**Knowledge of:**  
- Plan mode is designed for complex tasks involving large-scale changes, multiple valid
approaches, architectural decisions, and multi-file modifications  
- Direct execution is appropriate for simple, well-scoped changes (e.g., a

In [ ]:
query10 = "How to reduce API costs for automated analysis."
results = vectorstore.similarity_search(query10, k=5)
for i, result in enumerate(results):
    print(f"\n --- Result {i} ---\n")
    print(result.page_content)
    print(result.metadata)


 --- Result 0 ---

Section: Preparation Exercises > Exercise 4: Design and Debug a Multi-Agent Research Pipeline

**Objective: Practice orchestrating subagents, managing context passing, implementing error**
propagation, and handling synthesis with provenance tracking.  
**Steps:**  
1. Build a coordinator agent that delegates to at least two subagents (e.g., web search and  
document analysis). Ensure the coordinator's allowedTools includes "Task" and that
each subagent receives its research findings directly in its prompt rather than relying on
automatic context inheritance.  
2. Implement parallel subagent execution by having the coordinator emit multiple Task tool  
calls in a single response. Measure the latency improvement compared to sequential
execution.  
3. Design structured output for subagents that separates content from metadata: each  
finding should include a claim, evidence excerpt, source URL/document name, and
publication date. Verify that the synthesis subagent pres

## 5. BM25

#### Hitting a ceiling with breadcrumb on query2 because query keywords inside body not header, breadcrumb help when query vocabulary matches heading vocabulary, but not help when query terms in the body not the heading. First try BM25 to count term frequency to surface the exact phrase and boost the rank.

In [ ]:
#rank_bm25 (algorithm) already installed as the dependency for langchain BM25Retriever
from langchain_community.retrievers import BM25Retriever


In [ ]:
bm25_retriever = BM25Retriever.from_documents(md_chunks_breadcrumb)
bm25_retriever.k=10

results = bm25_retriever.invoke("What hands-on experience the idea candidate with for the exam?")
for i, doc in enumerate(results):
    print(f"Rank{i+1}: {doc.page_content}---\n")

Rank1: Section: Target Candidate Description

The ideal candidate for this certification is a solution architect who designs and implements
production applications with Claude. This candidate has hands-on experience with:  
- Building agentic applications using the Claude Agent SDK, including multi-agent
orchestration, subagent delegation, tool integration, and lifecycle hooks  
- Configuring and customizing Claude Code for team workflows using CLAUDE.md files,
Agent Skills, MCP server integrations, and plan mode  
- Designing Model Context Protocol (MCP) tool and resource interfaces for backend
system integration  
- Engineering prompts that produce reliable structured output, leveraging JSON schemas,
few-shot examples, and extraction patterns  
- Managing context windows effectively across long documents, multi-turn conversations,
and multi-agent handoffs  
- Integrating Claude into CI/CD pipelines for automated code review, test generation, and
pull request feedback---

Rank2: Secti

BM25Retrieval boosted up query2 rank 1 & 4, originally result 22 with partial answer with semantic search, that has "hands-on experience" in the chunk not heading. BM25 solved the specific vocabulary problem.

In [ ]:
result_Q7 = bm25_retriever.invoke("What topics are NOT covered in this exam?")

for i, doc in enumerate(result_Q7):
    print(f"Rank{i+1}: {doc.page_content}\n")

Rank1: Section: Preparation Exercises

Complete these hands-on exercises to build practical familiarity with the topics covered on the
exam. Each exercise is designed to reinforce knowledge across one or more exam domains.

Rank2: Section: Appendix > Out-of-Scope Topics

The following related topics will NOT appear on the exam:  
- Fine-tuning Claude models or training custom models  
- Claude API authentication, billing, or account management  
- Detailed implementation of specific programming languages or frameworks (beyond
what's needed for tool and schema configuration)  
- Deploying or hosting MCP servers (infrastructure, networking, container orchestration)  
- Claude's internal architecture, training process, or model weights  
- Constitutional AI, RLHF, or safety training methodologies  
- Embedding models or vector database implementation details  
- Computer use (browser automation, desktop interaction)  
- Vision/image analysis capabilities  
- Streaming API implementation o

BM25 boost Query7 to rank2(result1) from rank7(result6) with similarity_search,still missing 2 points.

In [ ]:
Query6 = "If I got 710, did I pass the exam?"
result_Q6 = bm25_retriever.invoke(Query6)
for i, doc in enumerate(result_Q6):
    print(f"Rank{i+1}: {doc.page_content}\n")

Rank1: Section: Sample Questions > Scenario: Claude Code for Continuous Integration

into smaller submissions of 3-4 files before the automated review runs. C) Switch to a
higher-tier model with a larger context window to give all 14 files adequate attention in one
pass. D) Run three independent review passes on the full PR and only flag issues that appear in at
least two of the three runs.  
**Correct Answer: A**  
Splitting reviews into focused passes directly addresses the root cause: attention dilution when
processing many files at once. File-by-file analysis ensures consistent depth, while a separate
integration pass catches cross-file issues. Option B shifts burden to developers without  
improving the system. Option C misunderstands that larger context windows don't solve
attention quality issues. Option D would actually suppress detection of real bugs by requiring
consensus on issues that may only be caught intermittently.

Rank2: Section: Sample Questions > Scenario: Claude Co

For query6 that has no shared vocabulary with the answer chunk(the chunk says "minimum passing score is 720, but query says"if I got 710"), BM25Retrieval push it to Rank3 from 1st with semantic/dense search. So use hybrid search for different queries 

## 6. Hybrid search to combine dense/semantic search(similar_search) + BM25 with reciprocal rank fusion

In [ ]:
from langchain.retrievers.ensemble import EnsembleRetriever

In [ ]:
dense_retriever = vectorstore.as_retriever(search_kwargs={"k": 10})

ensemble = EnsembleRetriever(
    retrievers = [dense_retriever,bm25_retriever],
    weights=[0.3,0.7]  # initialise equal weight between these two retrievers
)

for query in [
    "What hands-on experience the idea candidate with for the exam?",
    "What topics are NOT covered in this exam?",
]:
    print(f"\n ----{query} ----\n")
    for i, doc in enumerate(ensemble.invoke(query)):
        print(f"Rank{i+1}: {doc.page_content}\n")


 ----What hands-on experience the idea candidate with for the exam? ----

Rank1: Section: Preparation Exercises

Complete these hands-on exercises to build practical familiarity with the topics covered on the
exam. Each exercise is designed to reinforce knowledge across one or more exam domains.

Rank2: Section: Appendix > Exam Preparation Recommendations

validation-retry loops, design schemas with optional/nullable fields, and practice batch
processing with the Message Batches API.  
5. Practice prompt engineering techniques: Write few-shot examples for ambiguous  
scenarios. Define explicit review criteria to reduce false positives. Design multi-pass
review architectures for large code reviews.  
6. Study context management patterns: Practice extracting structured facts from verbose  
tool outputs, implementing scratchpad files for long sessions, and designing subagent
delegation to manage context limits.  
7. Review escalation and human-in-the-loop patterns: Understand when to esc

Hybrid search didnot solve the missing part for query7, changed retrieve k=10 to k=20 still missing. and tuned the weight to [0.3,0.7] (BM25-heavy) and [0.6,0.4] (dense-heavy) didnot make any difference to query7, but pushed query2 rank down to 21&22, will keep weight[0.5,0.5].

## 7. Reranker
#### See if rerank can solve query 7 missing 2 points issue

In [ ]:
from langchain.retrievers.document_compressors import CrossEncoderReranker
from langchain_community.cross_encoders import HuggingFaceCrossEncoder

In [ ]:
query7 = "What topics are NOT covered in this exam?"
cross_encoder = HuggingFaceCrossEncoder(model_name="BAAI/bge-reranker-base")
reranker = CrossEncoderReranker(model=cross_encoder, top_n=5)


In [ ]:
from langchain.retrievers import ContextualCompressionRetriever

compression_retriever = ContextualCompressionRetriever(
    base_compressor=reranker,
    base_retriever=ensemble
)

results = compression_retriever.invoke(query7)
for i, doc in enumerate(results):
    print(f"Rank{i+1}: {doc.page_content}\n")

Rank1: Section: Appendix > Out-of-Scope Topics

The following related topics will NOT appear on the exam:  
- Fine-tuning Claude models or training custom models  
- Claude API authentication, billing, or account management  
- Detailed implementation of specific programming languages or frameworks (beyond
what's needed for tool and schema configuration)  
- Deploying or hosting MCP servers (infrastructure, networking, container orchestration)  
- Claude's internal architecture, training process, or model weights  
- Constitutional AI, RLHF, or safety training methodologies  
- Embedding models or vector database implementation details  
- Computer use (browser automation, desktop interaction)  
- Vision/image analysis capabilities  
- Streaming API implementation or server-sent events  
- Rate limiting, quotas, or API pricing calculations  
- OAuth, API key rotation, or authentication protocol details  
- Specific cloud provider configurations (AWS, GCP, Azure)  
- Performance benchma

Reranker with top 20 boosted Rank2 (hybrid search) result up to Rank1 for query 7 Negation type of query - the same problem still exist, 2 points missing. The reranker worked and it correctly identified that rank2 was more relevant than rank1 from hybrid search for this query. So cross-encoder is doing its job, but the 2 missing points are still missing means it is not in the candidate pool. Tried top 50 chunks retrieve and it is still missing. It is vocabulary mismatch issue.

### 7.1 Evaluation with test set (60 Q&A pairs with ground_truth_chunk_ids) to convert manual eval to a reproducible eval.

In [ ]:

content_to_id = dict(zip(all_chunks["documents"], all_chunks["ids"]))


In [ ]:
K = 5

def ids_from_docs(docs):
    """Document objects - > chunk ids, skipping any text not found in the map."""
    return [content_to_id.get(d.page_content) for d in docs
            if content_to_id.get(d.page_content) is not None]

#baseline with dense_retriever(similarity_search)
def retrieve_dense(question):
    return ids_from_docs(dense_retriever.invoke(question))[:K]

def retrieve_bm25(question):
    return ids_from_docs(bm25_retriever.invoke(question))[:K]

def retrieve_hybrid(question):
    return ids_from_docs(ensemble.invoke(question))[:K]

def retrieve_reranked(question):
    return ids_from_docs(compression_retriever.invoke(question))[:K]


SYSTEMS = {
    "dense": retrieve_dense,
    "bm25": retrieve_bm25,
    "hybrid": retrieve_hybrid,
    "reranked": retrieve_reranked,
}


In [ ]:
# Recall@k
def recall_at_k(retrieved_ids:list, ground_truth_ids:list, k:int) -> float:
    return 1.0 if any(gid in retrieved_ids[:k] for gid in ground_truth_ids) else 0.0

In [ ]:
# Reciprocal Rank
def reciprocal_rank(retrieved_ids:list, ground_truth_ids: list) -> float:
    for rank, rid in enumerate(retrieved_ids, start=1):
        if rid in ground_truth_ids:
            return 1.0/rank
    return 0.0

In [ ]:
from collections import defaultdict
from statistics import mean

def evaluate(retrieve_fn, k=K):
    per_type = defaultdict(list)
    for row in rows:
        if row["type"] == "no_answer":
            continue
        retrieved = retrieve_fn(row["question"])
        rr = reciprocal_rank(retrieved, row["ground_truth_chunk_ids"])
        rk = recall_at_k(retrieved, row["ground_truth_chunk_ids"], k=K)
        per_type[row["type"]].append({"rr": rr, "rk": rk})
    return per_type

# run all four, collect overall numbers + keep per_type for drill-down
results = {name: evaluate(fn) for name, fn in SYSTEMS.items()}

# --- comparison table: one row per system ---
print(f"{'System':<10} {'MRR':>6} {'R@'+str(K):>6}")
for name, per_type in results.items():
    flat = [r for rs in per_type.values() for r in rs]
    mrr  = mean(r["rr"] for r in flat)
    rk   = mean(r["rk"] for r in flat)
    print(f"{name:<10} {mrr:>6.2f} {rk:>6.2f}")

System        MRR    R@5
dense        0.72   0.85
bm25         0.67   0.85
hybrid       0.79   0.92
reranked     0.87   0.94


In [ ]:
qtypes = sorted({row["type"] for row in rows if row["type"] != "no_answer"})
print(f"{'Type':<12} " + " ".join(f"{n:>10}" for n in SYSTEMS))
for qt in qtypes:
    cells = []
    for name in SYSTEMS:
        rs  = results[name][qt]
        mrr = mean(r["rr"] for r in rs)
        cells.append(f"{mrr:>10.2f}")
    print(f"{qt:<12} " + " ".join(cells))




Type              dense       bm25     hybrid   reranked
definition         0.83       0.79       0.83       0.78
enumeration        0.71       0.51       0.78       0.82
indirect           0.53       0.73       0.86       0.96
negation           0.79       0.65       0.69       0.92


In [ ]:
import json, datetime
from pathlib import Path

OUT = Path("eval/results_retrieval.jsonl")
run_date = datetime.date.today().isoformat()

def summarize(per_type):
    block, flat = {}, []
    for qt, rs in per_type.items():
        flat += rs
        block[qt] = {
            "n": len(rs),
            "mrr": mean(r["rr"] for r in rs),
            f"recall_at_{K}": mean(r["rk"] for r in rs),
            "raw": rs,
        }
    return block, flat

with OUT.open("a", encoding="utf-8") as f:
    for name, per_type in results.items():
        block, flat = summarize(per_type)
        f.write(json.dumps({
            "date": run_date,
            "system": name,
            "embedding_model": "sentence-transformers/all-MiniLM-L6-v2",
            "k": K,
            "n_questions": len(flat),
            "overall": {
                "mrr": mean(r["rr"] for r in flat),
                f"recall_at_{K}": mean(r["rk"] for r in flat),
            },
            "per_type": block,
        }) + "\n")

print(f"saved {len(results)} system rows -> {OUT}")


saved 4 system rows -> eval\results_retrieval.jsonl


## 8. Q & A
#### Select llama3.2:3b model as a starting pointto match my hardware - CPU system with 14GB RAM. Next qwen2.5:7b.

In [ ]:
from langchain_ollama import ChatOllama


In [89]:
llm = ChatOllama(
        model="qwen2.5:7b",
        temperature=0,
        num_predict=1024,
        num_ctx=8192
    )

def local_qa(query):

    docs = compression_retriever.invoke(query)
    
    context = "\n\n".join(
        f"[SOURCE {i+1}]:\n{doc.page_content}"
        for i, doc in enumerate(docs)
    )

    prompt = (
        f" Answer the question using only the provided context."
        f" If the question asks for a list, items, examples, or steps,"
        f" include EVERY relevant item from the context — do not summarise"
        f" or pick representatives. If the question asks for a specific"
        f" fact, give a focused answer."
        f" If the context contains only part of the answer, give the partial answer that the context does provide."
        f" Only if NONE of the sources contain any information relevant to the question should you reply exactly:Not found in context.\n\n"
        f"Sources:\n{context}\n\n"
        f"Question: {query}\n\n"
        f"Answer:\n"
    )

    response = llm.invoke(prompt)
    return {
        "result": response.content,
        "source_documents": docs,
    }


In [ ]:
query_dict = {
    "query2": "What hands-on experience the idea candidate with for the exam?",
    "query3": "What are out of scope topics?",
    "query7": "What topics are NOT covered in this exam?"
}

for key, query in query_dict.items():
    result = local_qa(query)
    print(f"\n------ {key}: {query} -------\n")
    print(f"Answer for {query}: {result['result']}\n\n")
    print("Source Documents:")
    for doc in result["source_documents"]:
        print(f"{doc.page_content}\n")



------ query2: What hands-on experience the idea candidate with for the exam? -------

Answer for What hands-on experience the idea candidate with for the exam?: - Building agentic applications using the Claude Agent SDK, including multi-agent orchestration, subagent delegation, tool integration, and lifecycle hooks  
- Configuring and customizing Claude Code for team workflows using CLAUDE.md files, Agent Skills, MCP server integrations, and plan mode  
- Designing Model Context Protocol (MCP) tool and resource interfaces for backend system integration  
- Engineering prompts that produce reliable structured output, leveraging JSON schemas, few-shot examples, and extraction patterns  
- Managing context windows effectively across long documents, multi-turn conversations, and multi-agent handoffs  
- Integrating Claude into CI/CD pipelines for automated code review, test generation, and pull request feedback  
- Making sound escalation and reliability decisions, including error handli

### 8.1 Run evals on generation

In [ ]:
import os, json

GEN_MODEL = "qwen2.5:7b"
GEN_PROMPT_V = "v2"
JUDGE_MODEL = "gemma2:9b"
JUDGE_PROMPT_V ="v1"

os.makedirs("eval/runs", exist_ok=True)
GEN_FILE = f"eval/runs/gen_{GEN_MODEL.replace(':','-')}_{GEN_PROMPT_V}.jsonl"
JUDGE_FILE = f"eval/runs/judge_{JUDGE_MODEL.replace(":","-")}_{JUDGE_PROMPT_V}_on_{os.path.basename(GEN_FILE)}"

def load_done_ids(path):
    if not os.path.exists(path):
        return set()
    return {json.loads(l)["id"] for l in open(path, encoding="utf-8")}

In [ ]:
# 1. generate all answers only use qwen2.5:7b
rows = [json.loads(l) for l in open("eval/eval_testset_v1.jsonl", encoding="utf-8")]

done = load_done_ids(GEN_FILE)

with open(GEN_FILE,"a",encoding="utf-8") as f:
    for row in rows:
        if row["id"] in done:
            continue
        rag = local_qa(row["question"])
        docs = rag["source_documents"]
        context = "\n\n" .join(
            f"[SOURCE {i+1}]:\n{doc.page_content}" for i, doc in enumerate(docs))
        f.write(json.dumps({
            "id": row["id"], "type":row["type"],"question": row["question"],
            "gold": row["answer"], "result": rag["result"], "context": context,
            "gen_prompt_v":GEN_PROMPT_V,
        }) + "\n")

        print("generated", row["id"])


In [80]:
from pydantic import BaseModel, Field
from langchain_ollama import ChatOllama

class JudgeScore(BaseModel):
    """Scores for one RAG answer, each 0.0-1.0."""
    faithfulness: float = Field(
        ge=0, le=1,
        description="Is every claim in the answer supported by the retrieved context?"
        "1.0 = fully grounded, 0.0 = hallucinated.")
    correctness: float = Field(
        ge=0, le=1,
        description="Does the answer match the ground-truth answer key?"
                    "1.0 = same facts, 0.0 = wrong.")
    completeness: float = Field(
        ge=0, le=1,
        description="Does it include ALL facts/items from the answer key?"
                    "Lower if it dropped items from a list.")
    reason: str = Field(description="One sentence justifying the scores.")

judge = ChatOllama(model=JUDGE_MODEL, temperature=0, num_predict=512, num_ctx=8192)
judge_structured = judge.with_structured_output(JudgeScore)

JUDGE_PROMPT = """You are grading a RAG system's answer.

QUESTION: {question}

GROUND-TRUTH ANSWER(the answer key) : {gold}

RETRIEVED CONTEXT the system saw:
{context}

SYSTEM'S ANSWER: {response}

Score faithfulness, correctness, and completeness.
Correctness and completeness are graded ONLY against the answer key.
Extra information beyond the answer key must NOT lower any score, as long as it supported by the context. 
The context is used only for faithfulness.
IMPORTANT: each score MUST be a decimal number between 0.0 and 1.0
(for example 0.0, 0.25, 0.5, 0.75, 1.0). Do NOT use a 1-5 or 1-10 scale.
1.0 means perfect, 0.0 means total failure.
Then give a one-sentence reason."""

def judge_answer(question, gold, response, context) -> JudgeScore:
    return judge_structured.invoke(JUDGE_PROMPT.format(
        question=question, gold=gold, response=response, context=context))


In [81]:

#2. judge all answers with gemma2:9b

ABSTAIN_PHRASE ="not found in context"

gen = [json.loads(l) for l in open(GEN_FILE, encoding="utf-8")]

done = load_done_ids(JUDGE_FILE)

with open(JUDGE_FILE, "a", encoding="utf-8") as f:
    for r in gen:
        if r["id"] in done:
            continue
        try:
            if r["type"] == "no_answer":
                abstained = ABSTAIN_PHRASE in r["result"].lower()
                scores = {
                    "faithfulness": 1.0,
                    "correctness": 1.0 if abstained else 0.1,
                    "completeness": 1.0,
                    "reason": "deterministic abstention check",
                }
            else:
                s = judge_answer(r["question"], r["gold"], r["result"], r["context"])
                scores = s.model_dump()
            f.write(json.dumps({
                "id": r["id"], 
                "type": r["type"],
                "judge_prompt_v": JUDGE_PROMPT_V,
                **scores,
            }) + "\n")
            f.flush()
            print("judged", r["id"], scores["faithfulness"], scores["correctness"], scores["completeness"])
        except Exception as e:
            print("⚠️ judge failed", r["id"], e)

judged Q01 1.0 1.0 1.0
judged Q02 1.0 1.0 1.0
judged Q03 0.2 0.0 0.0
judged Q04 1.0 1.0 1.0
judged Q05 1.0 1.0 1.0
judged Q06 1.0 1.0 1.0
judged Q07 1.0 1.0 1.0
judged Q08 1.0 1.0 1.0
judged Q09 1.0 1.0 1.0
judged Q10 1.0 1.0 1.0
judged Q11 1.0 1.0 1.0
judged Q12 0.8 0.9 0.9
judged Q13 1.0 1.0 1.0
judged Q14 0.6 0.0 0.0
judged Q15 1.0 1.0 1.0
judged Q16 1.0 1.0 1.0
judged Q17 1.0 1.0 1.0
judged Q18 1.0 1.0 1.0
judged Q19 1.0 1.0 1.0
judged Q20 1.0 1.0 1.0
judged Q21 1.0 1.0 1.0
judged Q22 1.0 1.0 1.0
judged Q23 0.9 1.0 1.0
judged Q24 1.0 1.0 1.0
judged Q25 1.0 1.0 1.0
judged Q26 1.0 1.0 1.0
judged Q27 0.9 1.0 1.0
judged Q28 0.9 1.0 0.8
judged Q29 0.6 0.0 0.0
judged Q30 0.9 1.0 1.0
judged Q31 0.9 1.0 1.0
judged Q32 1.0 1.0 1.0
judged Q33 1.0 1.0 1.0
judged Q34 0.8 0.9 0.9
judged Q35 1.0 1.0 1.0
judged Q36 1.0 1.0 1.0
judged Q37 1.0 1.0 1.0
judged Q38 0.6 0.0 0.0
judged Q39 0.6 1.0 1.0
judged Q40 1.0 1.0 1.0
judged Q41 0.8 1.0 1.0
judged Q42 0.8 0.0 1.0
judged Q43 1.0 1.0 1.0
judged Q44 